### ISSUE TITLE
Device status not updated during camera triggers

### ISSUE THREAD

**Issue by aeirola** [timrogers/litra-autotoggle#10]
Hi again, sorry for constantly nagging about supporting my strange laptop hot-plugging setup, but noticed some issues in the implementation of #3. Really appreciate the work, and you have already done more than enough to support my use-cases, so don't sweat it if it is hard to find the time to handle these cases in the way I was hoping. Just thought to report in case it was something you happened to overlook in the implementation.

Anyways, it appears as if the device status is not updated when the camera is turned on or off. This causes a couple of unexpected behaviours:

## Connecting light after startup

1. Disconnect lights
2. Start `litra-autotoggle`
3. Connect lights
4. Start camera
5. Light does not turn on

This produces the following log output

```
% litra-autotoggle                   
No Litra devices found
Starting `log` process to listen for video device events...
Listening for video device events...
// Connect light to computer
Detected that a video device has been turned on, attempting to turn on Litra device...
No Litra devices found
```

Would have expected litra-autotoggle to scan for new devices when it detects that the camera has turned on, and then turn on the newly connected devices.

## Disconnecting light after startup

1. Connect lights
2. Start `litra-autotoggle`
3. Disconnect lights
4. Start camera
5. Light does not turn on, and `litra-autotoggle` process crashes

```
% litra-autotoggle
Found Litra Glow device (serial number: 2141FE000P88)
Starting `log` process to listen for video device events...
Listening for video device events...
// Disconnect light
Detected that a video device has been turned on, attempting to turn on Litra device...
HID error occurred: hidapi error: hid_open_path: device mach entry not found with the given path
```

Would expect the process to keep running even though the devices are not available anymore and treat the situation as if no devices are connected.

## Environment

- litra-autotoggle version: 0.3.0
- Operating system: macOS 15.0.1
- Device: Litra Glow

**Comment by timrogers** [comment]
What you describe is what should be happening 😓 I will take a look when I have some time!

**Comment by timrogers** [comment]
Can you run `litra-autotoggle --version`, just to double check for any version surprises? 😓

**Comment by aeirola** [comment]
Okay, so I'm not a Rust developer, but I think I managed to trace the issue. It goes something like this:

1. When `litra-autotoggle` notices a change in camera status, it fetches the list of supported devices using the `litra-rs` library at https://github.com/timrogers/litra-autotoggle/blob/v0.4.0/src/main.rs#L108
2. In the `litra-rs` library, this fetches the list of devices using `HidApi` at https://github.com/timrogers/litra-rs/blob/v2.1.0/src/lib.rs#L61
3. The [`device_list`](https://docs.rs/hidapi/latest/hidapi/struct.HidApi.html#method.device_list) method in `HidApi` seems to use cached results as there are no calls to [`refresh_devices`](https://docs.rs/hidapi/latest/hidapi/struct.HidApi.html#method.refresh_devices), and the same instance of `HidApi` that was created on litra-autotoggle start is reused.

So I think the solution would be to at some point after a camera change is identified force the HidApi to refresh the internal list of devices before attempting to turn the lights on or off.

Sorry if my terminology is not consistent with Rust developer community conventions. I hope the message is understandable nonetheless.

**Comment by timrogers** [comment]
Totally makes sense! I am travelling at the moment without my Litra, so it's hard for me to fix this, but I will pick it up when I get back.

**Comment by timrogers** [comment]
This should be fixed in v0.4.1! Can you give it another go?

**Comment by aeirola** [comment]
Yes, works flawlessly now! Thanks for taking the time to fix it.
